javatests/arcs/android/entity/TtlHandleTest.kt

```diff
@@ -43,137 +56,286 @@ class TtlHandleTest {
+                assertThat(handle2.fetch()).isEqualTo(null)
 
-        assertThat(handle2.fetch()).isEqualTo(null)
-        
-        val entity2 = DummyEntity().apply {
-            num = 2.0
-            texts = setOf("2", "two")
+                log("Replacing with entity 2 in handle 2.")
+                modifications.add(handle2.store(entity2))
+                assertThat(handle2.fetch()).isEqualTo(entity2)
+
+                log("Clearing handle 2")
+                modifications.add(handle2.clear())
+                assertThat(handle2.fetch()).isNull()
+                readyJob.complete()
+            } catch (e: Throwable) {
+                readyJob.completeExceptionally(e)
+            }
         }
-        handle2.store(entity2)
-        assertThat(handle2.fetch()).isEqualTo(entity2)
-        handle2.clear()
-        assertThat(handle2.fetch()).isNull()
+        readyJob.join()
+        modifications.joinAll()
+        log("all done")
     }
 
     @Test
-    fun collectionWithExpiredEntities() = runBlockingTest {
+    fun collectionWithExpiredEntities() = runBlocking {
         val handle = createCollectionHandle()
         val entity1 = DummyEntity().apply {
             num = 1.0
             texts = setOf("1", "one")
         }
-        // Set a time in the past. So that this entity is already expired.
-        fakeTime.millis = 1L
-        handle.store(entity1)
-        // Then set time to now.
-        fakeTime.millis = System.currentTimeMillis()
-
         val entity2 = DummyEntity().apply {
             num = 2.0
             texts = setOf("2", "two")
         }
-        handle.store(entity2)
-
-        // Entity1 is already expired.
-        assertThat(handle.fetchAll()).containsExactly(entity2)
-        
-        // Simulate periodic job triggering.
-        databaseManager.removeExpiredEntities()
-
-        assertThat(handle.fetchAll()).containsExactly(entity2)
-
         val entity3 = DummyEntity().apply {
             num = 3.0
             texts = setOf("3", "three")
         }
-        handle.store(entity3)
-        assertThat(handle.fetchAll()).containsExactly(entity2, entity3)
+        val entity4 = DummyEntity().apply {
+            num = 4.0
+            texts = setOf("4", "four")
+        }
+
+        var readyJob = Job()
+        val activeWrites = mutableListOf<Job>()
+        handle.onReady {
+            log("handle.onReady called.")
+            // Set a time in the past. So that this entity is already expired.
+            fakeTime.millis = 1L
+            log("handle.store(entity1)")
+            activeWrites.add(handle.store(entity1))
+
+            // Then set time to now and add another entity - not expired.
+            fakeTime.millis = System.currentTimeMillis()
+            log("handle.store(entity2)")
+            activeWrites.add(handle.store(entity2))
+
+            try {
+                // TODO(b/152361041): after expired entities are filtered out on read, this should
```

Comment:
just noticing this now, b/152361041 was already fixed at the time of this pr, only entity2 is in the handle.
completeExceptionally and join don't work as you think: join() is happy even if the job completed exceptionally (see https://github.com/Kotlin/kotlinx.coroutines/issues/1527).

For example try adding a assertThat(true).isFalse() inside this try block, and observe the test still passing.

This means every time we are using try {} catch { completeExceptionally } we are not really checking those assertThat